File: src/main/java/frc/robot/base/input/Controller.java
package frc.robot.base.input;

import java.util.HashMap;

import edu.wpi.first.wpilibj.Joystick;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class Controller {

    private Joystick joystick;
    private HashMap<Integer, Boolean> buttonsPressed;

    public Controller(Joystick joystick) {
        this.joystick = joystick;
        this.buttonsPressed = new HashMap<>();
        for (Button button : Button.values()) {
            int id = button.getId();
            buttonsPressed.put(id, false);
        }
    }

    /**
     * @param button the button you want to specify
     * @return whether the specified button is current being pressed
     */
    public boolean buttonDown(Button button) {
        return joystick.getRawButton(button.getId());
    }

    /**
     * Returns true during the first loop the specified button is pressed
     * 
     * @param button the button you want to specify
     * @return whether the specified button was just pressed
     */
    public boolean buttonPressed(Button button) {
        return (!buttonsPressed.get(button.getId()) && joystick.getRawButton(button.getId()));
    }

    /**
     * Returns the value of an axis on the controller
     * 
     * @param axis the axis you want to measure
     * @return the value of the axis
     */
    public double getAxis(Axis axis) {
        return joystick.getRawAxis(axis.getId());
    }

    /**
     * Returns the value of a pov on the controller
     * 
     * @param axis the pov you want to measure
     * @return the angle of the POV in degrees, or -1 if the POV is not pressed.
     */
    public int getPov(Pov pov) {
        return joystick.getPOV(pov.getId());
    }

    /**
     * A method that should be run after the main periodic code;
     * makes buttonPressed(Button button) work
     */
    public void postPeriodic() {
        for (Button button : Button.values()) {
            int id = button.getId();
            boolean pressed = joystick.getRawButton(id);
            buttonsPressed.put(id, pressed);
            SmartDashboard.putBoolean("vars2/controllers/port_" + joystick.getPort() + "/buttons/" + button, pressed);
        }
    }

}

File: src/main/java/frc/robot/input/Controller.kt
package frc.robot.input

import edu.wpi.first.wpilibj.Joystick
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import java.util.*

class Controller(private val joystick: Joystick) {
    private val buttonsPressed: HashMap<Int, Boolean> = HashMap()
    /**
     * @param button - the button you want to specify
     * @return whether the specified button is current being pressed
     */
    fun buttonDown(button: Button): Boolean {
        return joystick.getRawButton(button.id)
    }

    /**
     * Returns true as soon as the specified button is pressed and then goes back to
     * false until it's pressed again
     *
     * @param button - the button you want to specify
     * @return whether the specified button was just pressed
     */
    fun buttonPressed(button: Button): Boolean {
        return !buttonsPressed[button.id]!! && joystick.getRawButton(button.id)
    }

    /**
     * Returns the value of an axis on the controller
     *
     * @param axis - the axis you want to measure
     * @return the values of the axis
     */
    fun getAxis(axis: Axis): Double {
        return joystick.getRawAxis(axis.id)
    }

    /**
     * A method that should be run after the main periodic code; makes
     * buttonPressed(Button button) work
     */
    fun postPeriodic() {
        for (button in Button.values()) {
            val id = button.id
            val pressed = joystick.getRawButton(id)
            buttonsPressed[id] = pressed
            SmartDashboard.putBoolean("vars2/controllers/port_" + joystick.port + "/buttons/" + button, pressed)
        }
    }

    init {
        for (button in Button.values()) {
            val id = button.id
            buttonsPressed[id] = false
        }
    }
}